api/src/main/java/org/apache/iceberg/io/SupportsPrefixOperations.java

```diff
@@ -0,0 +1,53 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.io;
+
+import java.util.stream.Stream;
+
+/**
+ * This interface is intended as an extension for FileIO implementations
+ * to provide additional prefix based operations that may be useful in
+ * performing supporting operations.
+ */
+public interface SupportsPrefixOperations {
+
+  /**
+   * Return a stream of all files under a prefix.
+   *
+   * Hierarchical file systems (e.g. HDFS) may impose additional restrictions
+   * like the prefix mush fully match a directory whereas key/value object
+   * stores may allow for arbitrary prefixes.
+   *
+   * @param prefix prefix to list
+   * @return stream of file information
+   */
+  Stream<FileInfo> listPrefix(String prefix);
```

Comment:
We don't usually expose `Stream` in our APIs. Normally this would be an `Iterable<FileInfo>`. Is there a reason to use a stream instead?

Reply:
We probably shouldn't expose `Iterable` here because this should really be "traversable once" and it gets overly complicated to fully support the contract.  Hadoop uses `Iterator`, but has its own implementation as you can see from the adaptor.  If your goal is to simply keep streams out of the API, we should probably fallback to `Iterator`.